src/http/modules/ngx_http_upstream_least_conn_module.c

```diff
@@ -278,8 +285,18 @@ ngx_http_upstream_get_least_conn_peer(ngx_peer_connection_t *pc, void *data)
         }
 
         ngx_http_upstream_rr_peers_wlock(peers);
+
+#if (NGX_HTTP_UPSTREAM_ZONE)
+        if (peers->config && rrp->config != *peers->config) {
+            goto busy;
+        }
+#endif
```

Comment:
This block makes no sense, since `busy` is right next to it. Also, lock/unlock can be avoided as well. We can just call `ngx_http_upstream_get_least_conn_peer()` recursively and return its result. The nested call will assign the same `pc->name`.

```
@@ -103,7 +103,7 @@ ngx_http_upstream_get_least_conn_peer(ngx_peer_connection_t *pc, void *data)
 
     time_t                         now;
     uintptr_t                      m;
-    ngx_int_t                      rc, total;
+    ngx_int_t                      total;
     ngx_uint_t                     i, n, p, many;
     ngx_http_upstream_rr_peer_t   *peer, *best;
     ngx_http_upstream_rr_peers_t  *peers;
@@ -278,19 +278,7 @@ failed:
 
         ngx_http_upstream_rr_peers_unlock(peers);
 
-        rc = ngx_http_upstream_get_least_conn_peer(pc, rrp);
-
-        if (rc != NGX_BUSY) {
-            return rc;
-        }
-
-        ngx_http_upstream_rr_peers_wlock(peers);
-
-#if (NGX_HTTP_UPSTREAM_ZONE)
-        if (peers->config && rrp->config != *peers->config) {
-            goto busy;
-        }
-#endif
+        return ngx_http_upstream_get_least_conn_peer(pc, rrp);
     }
 
 #if (NGX_HTTP_UPSTREAM_ZONE)

```

Reply:
I could swear I dropped this goto busy after your review in July. Anyways, fixed. Thanks for catching!

Regarding the lock/unlock optimization: it is a preexisting issue introduced to reduce the diff with our other repository. I can submit another PR addressing that, but I believe it would only complicate merges.